include/mapbox/variant.hpp

```diff
@@ -861,6 +862,22 @@ class variant
         return detail::binary_dispatcher<F, V, R, Types...>::apply(v0, v1, std::forward<F>(f));
     }
 
+    // match
+    // unary
+    template <typename... Fs>
+    auto VARIANT_INLINE match(Fs&&... fs) const
+        -> decltype(detail::dispatcher<typename ::mapbox::util::visitor<Fs...>::type, variant, typename detail::result_of_unary_visit<typename ::mapbox::util::visitor<Fs...>::type, first_type>::type, Types...>::apply_const(*this, ::mapbox::util::make_visitor(std::forward<Fs>(fs)...)))
+    {
+        return detail::dispatcher<typename ::mapbox::util::visitor<Fs...>::type, variant, typename detail::result_of_unary_visit<typename ::mapbox::util::visitor<Fs...>::type, first_type>::type, Types...>::apply_const(*this, ::mapbox::util::make_visitor(std::forward<Fs>(fs)...));
+    }
+    // non-const
+    template <typename... Fs>
+    auto VARIANT_INLINE match(Fs&&... fs)
+        -> decltype(detail::dispatcher<typename ::mapbox::util::visitor<Fs...>::type, variant, typename detail::result_of_unary_visit<typename ::mapbox::util::visitor<Fs...>::type, first_type>::type, Types...>::apply(*this, ::mapbox::util::make_visitor(std::forward<Fs>(fs)...)))
+    {
+        return detail::dispatcher<typename ::mapbox::util::visitor<Fs...>::type, variant, typename detail::result_of_unary_visit<typename ::mapbox::util::visitor<Fs...>::type, first_type>::type, Types...>::apply(*this, ::mapbox::util::make_visitor(std::forward<Fs>(fs)...));
+    }
```

Comment:
This got seriously out of hand - not sure what to do about it, though:
- We need the `auto fn .-> decltype(expr) { return expr; }` pattern to support C++11, therefore we have to repeat `expr` twice. Once for the function's body, once to deduce the function's return type.
- All the expression does is `return dispatcher<Lambdas, Self, ReturnType>::apply(*this, lambdas);` - we could introduce type aliases in the function body but the `decltype()` part still has to be a single expression